### PR TITLE
Allow to provide custom headers in requests

### PR DESCRIPTION
**Pull request by pablog**
## Summary ##

Fixes https://github.com/Wolox/wolmo-networking-ios/issues/8

Allow to provide custom headers in requests in `AbstractRepository`.

### CODE REVIEW

**Comment by woloxci** [comment]
**CodeStats metrics details:**

:white_check_mark: Code Coverage (Slather): 82.48


Click [here](http://codestats.wolox.com.ar/organizations/wolox/projects/wolmo-networking-ios/branches/allow-headers) for more details.

**Comment by woloxci** [comment]
**CodeStats metrics details:**

:white_check_mark: Code Coverage (Slather): 82.48
:white_check_mark: Code Quality (Swiftlint): 94.0


Click [here](http://codestats.wolox.com.ar/organizations/wolox/projects/wolmo-networking-ios/branches/allow-headers) for more details.

**Comment by woloxci** [comment]
**CodeStats metrics details:**

:white_check_mark: Code Coverage (Slather): 81.09
:white_check_mark: Code Quality (Swiftlint): 94.0


Click [here](http://codestats.wolox.com.ar/organizations/wolox/projects/wolmo-networking-ios/branches/allow-headers) for more details.

File: Networking/Repository/AbstractRepository.swift [hunk 1]
```diff
@@ -225,17 +238,25 @@ fileprivate extension AbstractRepository {
...
+        headers: [String: String]? = .none,
+        requiresSession: Bool) -> SignalProducer<RawDataResponse, RepositoryError> {
         guard let url = buildURL(path: path) else { return SignalProducer(error: .invalidURL) }
         
+        var headers = headers
+        if requiresSession {
+            guard _sessionManager.isLoggedIn else { return SignalProducer(error: .unauthenticatedSession) }
+            headers = headers ?? [:]
+            headers![authenticationHeaders.first!.key] = authenticationHeaders.first!.value
+        }
```
Comment (danielaRiesgo) [inline Networking/Repository/AbstractRepository.swift]: Sorry I don't get this too much. Couldn't you just add the authenticationHeaders to the normal headers and use that? Something like `headers = (headers ?? [:]) + authenticationHeaders`.
Btw, I would add in the documentation somewhere (maybe in the headers parameter in each function or in the AbstractRepository class description) that the authentication headers don't need to be passed, the abstract repository will handle them automatically with the session service, that "headers" is just for extra non-authentication headers.

**Comment by danielaRiesgo** [comment]
And couldn't we make the AbstractRepository be created with default headers that should go with every request (like for example de Content-Type and Accept headers) so that in the child repository you don't have to add those default headers to the normal headers in every all to `perform` methods.

File: Networking/Repository/AbstractRepository.swift (hunk 1 above)
Comment (pablog) [inline Networking/Repository/AbstractRepository.swift]: Sure, you mean create a `Dictionary` extension and add the `+` operator?

I'll add the comment.

**Comment by pablog** [comment]
Sure, I'll do that too.

File: Networking/Repository/AbstractRepository.swift (hunk 1 above)
Comment (danielaRiesgo) [inline Networking/Repository/AbstractRepository.swift]: You can copy Core's: https://github.com/Wolox/wolmo-core-ios/blob/master/Core/Extensions/Foundation/Dictionary.swift

**Comment by woloxci** [comment]
**CodeStats metrics details:**

:white_check_mark: Code Coverage (Slather): 82.07
:white_check_mark: Code Quality (Swiftlint): 94.0


Click [here](http://codestats.wolox.com.ar/organizations/wolox/projects/wolmo-networking-ios/branches/allow-headers) for more details.

**Comment by pablog** [comment]
I could modify the function `append`:

```
mutating func append(contentsOf dictionary: [Key: Value]) {
        for (key, value) in dictionary where self[key] == nil {
            updateValue(value, forKey: key)
        }
    }
```

This way it keep the value provided in headers, instead of overriding it with the default values in `AbstractRepository`, just in case I want to provide default values for these properties.

What do you think?

**Comment by danielaRiesgo** [comment]
Hmm, sorry but I don't like that very much. I prefer things to be explicit. The authentication header comes from the session service you give the abstract repository when created. But then you are automatically added json headers just because. And if you want for one repository xml instead of json you would need to add them to your headers every time before calling the perform function, and that's what we are trying to avoid by having this customizable default headers. I don't think it's very customizable like this. If not, the default header could be a var that could be overwritten by the child. (I prefer the init just to be more explicit, but it would be ok none the less.)

**Comment by woloxci** [comment]
**CodeStats metrics details:**

:white_check_mark: Code Coverage (Slather): 81.35
:white_check_mark: Code Quality (Swiftlint): 94.0


Click [here](http://codestats.wolox.com.ar/organizations/wolox/projects/wolmo-networking-ios/branches/allow-headers) for more details.

File: Networking/Repository/AbstractRepository.swift [hunk 2]
```diff
@@ -133,18 +133,25 @@ open class AbstractRepository {
     fileprivate let _sessionManager: SessionManagerType
     fileprivate let _requestExecutor: RequestExecutorType
     
+    fileprivate let _defaultHeaders: [String: String]?
```
Comment (danielaRiesgo) [inline Networking/Repository/AbstractRepository.swift]: Does it have to be `[String: String]` ? Isn't only `[String: Any]` needed?

File: Networking/Repository/AbstractRepository.swift [hunk 3]
```diff
@@ -133,18 +133,25 @@ open class AbstractRepository {
...
         _requestExecutor = requestExecutor
         _sessionManager = sessionManager
+        _defaultHeaders = defaultHeaders
     }
     
-    public init(networkingConfiguration: NetworkingConfiguration, sessionManager: SessionManagerType) {
+    public init(networkingConfiguration: NetworkingConfiguration,
+                sessionManager: SessionManagerType,
+                defaultHeaders: [String: String]? = .none) {
         _networkingConfiguration = networkingConfiguration
```
Comment (danielaRiesgo) [inline Networking/Repository/AbstractRepository.swift]: Shouldn't this init call the upper one with `requestExecutor ` argument in `defaultRequestExecutor(networkingConfiguration: networkingConfiguration)` ?

File: Networking/Repository/AbstractRepository.swift (hunk 2 above)
Comment (pablog) [inline Networking/Repository/AbstractRepository.swift]: It expects `[String: String]`: 

`Alamofire` defines: `public typealias HTTPHeaders = [String : String]`

File: Networking/Repository/AbstractRepository.swift (hunk 3 above)
Comment (pablog) [inline Networking/Repository/AbstractRepository.swift]: Sure. Fixed.

**Comment by woloxci** [comment]
**CodeStats metrics details:**

:white_check_mark: Code Coverage (Slather): 81.35
:white_check_mark: Code Quality (Swiftlint): 94.0


Click [here](http://codestats.wolox.com.ar/organizations/wolox/projects/wolmo-networking-ios/branches/allow-headers) for more details.

**Pull request merged**